Add generic functionality

dmontagu opened the pull request:
## Change Summary

Adds some functionality around generics.

There are definitely some edge cases not perfectly handled yet, but I wanted to get your take on this before I invest more effort. A number of people from the FastAPI community seem interested in this functionality, but if you think it's going to be too hard to maintain then we can let it go.

For what it's worth, I think the `create_model` function is actually doing 90% of the work necessary to get this working.

## Related issue number

https://github.com/samuelcolvin/pydantic/issues/556

## Checklist

* [x] Unit tests for the changes exist
* [x] Tests pass on CI and coverage remains at 100%
* [x] Documentation reflects the changes where applicable
* [x] `HISTORY.rst` has been updated
  * if this is the first change since a release, please add a new section
  * include the issue number or this pull request number `#<number>`
  * include your github username `@<whomever>`

dmontagu commented:
I think it would be hard to get this working in python 3.6 due to differences in the way generics work. But I could look into it if necessary.

samuelcolvin commented:
I'll look at this futher, but overall looks cool.

I don't think 3.6 matters, just 3.7 support for this is fine. 

Please use single quotes, I'll do my best to add linting for single quotes somehow soon.

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -0,0 +1,67 @@
+        if not isinstance(params, tuple):
+            params = (params,)
+        generic_alias = super().__class_getitem__(params)
+        __parameters__ = generic_alias.__origin__.__parameters__
```
no need for `__*__` in local variables.

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -0,0 +1,67 @@
+        validators = gather_validators(cls)
+        type_hints = get_type_hints(cls)
+        new_type_hints = resolve_generic_type_hints(type_hints, typevars_map)
+        fields = {}
```
```suggestion
        fields = {k: (v, getattr(cls, k, ...)) for k, v in new_type_hints.items()}
```

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -0,0 +1,67 @@
+
+
+def resolve_generic_type_hints(generic_type_hints: Dict[str, Any], typevars_map: Dict[Any, Any]) -> Dict[str, Any]:
+    new_type_hints = {}
```
again comprehension.

samuelcolvin commented on tests/test_generics.py:
```diff
@@ -0,0 +1,113 @@
+from pydantic.generics import BaseGenericModel, generic_name
+
+
+skip_not_37 = pytest.mark.skipif(not sys.version_info >= (3, 7), reason='generics only supported for python 3.7')
```
```suggestion
skip_36 = pytest.mark.skipif(not sys.version_info >= (3, 7), reason='generics only supported for python 3.7 and above')
```

samuelcolvin commented on tests/test_generics.py:
```diff
@@ -0,0 +1,113 @@
+    class Result(BaseGenericModel, Generic[data_type]):
+        data: data_type
+
+        class Config:
```
not required I think.

samuelcolvin commented on tests/test_generics.py:
```diff
@@ -0,0 +1,113 @@
+        class Config:
+            allow_mutation = False
+
+    result = Result[int](data=1)
```
```suggestion
    result = Result[int](data=1)
    assert result.data == 1
```

samuelcolvin commented on pydantic/generics.py:
```diff
@@ -0,0 +1,67 @@
+
+class BaseGenericModel:
+    """
+    This might benefit from being modeled more closely on something like typing_extensions.Protocol
```
This is more a comment than a docstring, maybe you could move this to a comment or into the PR and add a docstring.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,76 @@
+from pydantic.class_validators import gather_validators
+
+
+class GenericModel:
```
I figured GenericModel was an appropriate name but I can change it back if desired.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,76 @@
+    def __class_getitem__(cls, params: Union[Any, Tuple[Any, ...]]) -> Type[BaseModel]:
+        if not isinstance(params, tuple):
+            params = (params,)
+        if cls is GenericModel:
```
Reworked so that you never have to inherit separately from both BaseGenericModel and Generic.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,76 @@
+        if cls is GenericModel:
+            if not all(isinstance(x, TypeVar) for x in params):
+                raise TypeError(f'Each parameter to {cls.__name__} must be a TypeVar')
+            return type('ParameterizedGenericModel', (GenericModel,), {'__parameters__': params})
```
I initially did this by declaring ParameterizedGenericModel as a class; this caused mypy to crash.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,76 @@
+        def __init__(self, *args: Any, **kwargs: Any) -> None:
+            ...
+
+    @no_type_check
```
lru_cache is causing errors here; I'm working on a caching solution

dmontagu replied:
A dict worked

dmontagu commented on tests/test_generics.py:
```diff
@@ -0,0 +1,206 @@
+    S = TypeVar('S')
+    R = TypeVar('R')
+
+    class OuterModel(GenericModel[T, S, R]):
```
Note the new API

dmontagu commented on setup.py:
```diff
@@ -57,6 +57,7 @@ def extra(self):
         os.environ['CFLAGS'] = '-O3'
         ext_modules = cythonize(
             'pydantic/*.py',
+            exclude=['pydantic/generics.py'],
```
https://github.com/cython/cython/issues/2753

drunkwcodes commented:
Can it be some constraints on `Generics`, e.g. "not iterable"?
It is less expressive without type constraints and losts the purpose with type constraints.

dmontagu commented:
> Can it be some constraints on `Generics`, e.g. "not iterable"?
> It is less expressive without type constraints and losts the purpose with type constraints.

I'm not sure exactly what you mean. Are you suggesting support for something like `typing_extensions.Protocol`?

drunkwcodes commented:
@dmontagu It is useful when you want to construct a list with `*args` and ensure there is no iterable in it.

dmontagu commented:
Hmm, I didn't notice this because mypy doesn't run over the tests by default, but it looks mypy doesn't like type variables if you don't explicitly inherit from Generic (or another `typing` class). So it looks like it may be necessary to change back to
```python
class Model(GenericModel, Generic[T, S]):
    ...
```

dmontagu commented:
Okay, everything worked out nicer than expected:
* I was able to inherit from `BaseModel` without any problems, which helps mypy with all the built-in BaseModel methods.
* If we want to support differing numbers of type variables (which I think we should), I think it's *necessary* that generic classes also inherit from `Generic[T]` or similar. This appears to be imposed by `typing` since you can't create your own custom variadic generic type yet. (This is an open issue in python/typing.)

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,87 @@
+        created_model = create_model(
+            model_name=model_name,
+            __module__=cls.__module__,
+            __base__=None,
```
I think I'm not enough of a pydantic pro to know whether it would ever make sense to specify `__base__` here. It seems much more likely that `__config__` would be useful. (I am explicitly specifying `__base__` here to get mypy to stop giving errors about the type of `fields`, which it strangely does if you skip an argument.)

samuelcolvin replied:
looks fine to me, as your test shows `Config` works which I think is sufficient.

drunkwcodes commented:
@dmontagu `typing` works sometimes. It depends.

drunkwcodes commented on pydantic/generics.py:
```diff
@@ -0,0 +1,81 @@
+    __slots__ = ()
+
+    def __new__(cls, *args: Any, **kwargs: Any) -> NoReturn:
+        if cls is GenericModel:
```
Does it mean it is not abstract and still can not be initialized without subclassing and not useful without implenting the function by users themselves?

samuelcolvin commented:
Looks good to me, please can you:
* update history
* add a section to docs
* "Hmm, I didn't notice this because mypy doesn't run over the tests by default..." add to the mypy external tests (`tests/mypy_test_success.py`) to check that `Generic` works with mypy

@drunkwcodes I'm afraid I don't understand what you're asking/concerned about. Probably best to give a code example to demonstrate your point. If you're not sure how something will work, it would be good if you checked out this PR and ran code yourself to check, then reported back here if you've found a problem.

drunkwcodes commented:
@samuelcolvin The `test_generic()` itself is expressive enough to show how a type checking function can be faked for other implementers.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,82 @@
+            __validators__=validators,
+            **fields,
+        )
+        created_model.Config = cls.Config
```
I got everything working now so that methods defined on the generic are inherited properly, etc. (See tests for more clarity). The `create_model` function doesn't let you specify both a config and a base class though, so I resorted to manually setting it afterward.

Since `created_model` is a concrete instantiation, I think it makes sense to have all subclasses actually share the *same* config instance, but let me know if you think it would be better to create a copy.

tiangolo commented:
Great job @dmontagu! Impressive 👏👏🎉

Even the docs are great! 📝🚀

dmontagu commented:
One thing that might be problematic is that I am placing `[` and `]` in the name of the created model. This makes all the associated validation messages cleaner / easier to parse, but I could see it causing unforeseen issues with the schema-related features due to having special characters (since you won't get special characters in the name of any model declared e.g. by inheriting from `BaseModel`).

In particular, @tiangolo I could see this causing issues with the OpenAPI spec generated by FastAPI. Is there anything to worry about here? (Maybe there's a config setting that let's you override the name of the model used in the spec? I'm a little out of my depth here.)

dmontagu commented:
Okay, I tested it out, and there don't seem to be any problems with FastAPI. However, the generated names (which currently may include `[`, `]`, ` `, `.`, and `,` violate the openapi spec, and prevent the use of `openapi-generator` because they don't meet the regex `^[a-zA-Z0-9\.\-_]+$`.

dmontagu commented on pydantic/schema.py:
```diff
@@ -346,6 +347,7 @@ def get_model_name_map(unique_models: Set[Type['BaseModel']]) -> Dict[Type['Base
     conflicting_names: Set[str] = set()
     for model in unique_models:
         model_name = model.__name__
+        model_name = re.sub(r"[^a-zA-Z0-9.\-_]", "_", model_name)
```
@samuelcolvin I expect this line may need to be removed / changed before merging, but adding this regex substitution fixes all problems with generic class names for `openapi-generator` (and I imagine for most json-schema-based tools). As far as I can tell, it isn't possible to generate a name that would be modified by this regex substitution in the normal course of pydantic usage (i.e., inheriting from `BaseModel`), including them is a violation of the open api spec (and presumably others), and I am having a hard time imagining a scenario where you would want those characters in your model name anyway.

From what I can tell, this substitution would also have little-to-no impact on performance given it's only involved in schema generation and is a fairly cheap regex.

I'm reluctant to change the naming scheme for generic classes because the `repr` and error messages are so clear. At the same time, I think json schemas are one of the most natural contexts for the use of generics (and my intended use is specifically with openapi / `openapi-generator`), so I think it would be unfortunate if they were slightly broken for this use case.

I'd be interested in your thoughts.

tiangolo replied:
LGTM

This function is mainly/only used to generate the JSON Schemas (and it's reused by FastAPI).

I think it would be safe/ideal to apply this. And I see you are using the same regex in the OpenAPI spec :heavy_check_mark:

dmontagu commented on pydantic/generics.py:
```diff
@@ -56,6 +56,8 @@ def __class_getitem__(  # type: ignore
         _generic_types_cache[(cls, params)] = created_model
+        if len(params) == 1:
+            _generic_types_cache[(cls, params[0])] = created_model
         return created_model
```
You need to have the non-tuple version in the cache so cache-access can be the first line of the function (rather than tuple-making), ensuring minimum overhead. I think it makes sense to also add the tuple version to the cache to minimize the amount of special cases (and after all it is possible to pass a single-element tuple to the `__class_getitem__` function).

dmontagu commented on tests/test_generics.py:
```diff
@@ -131,7 +143,7 @@ class Config:
     with pytest.raises(TypeError) as exc_info:
         GenericModel()
-    assert str(exc_info.value) == 'Type GenericModel cannot be instantiated; it can be used only as a base class'
+    assert str(exc_info.value) == 'Type GenericModel cannot be used without generic parameters, e.g. GenericModel[T]'
```
This might be slightly confusing because you also have to inherit from Generic before this will work, but if you try (e.g. by doing GenericModel[int]) it will then tell you to inherit from typing.Generic. So I think it's okay.

dmontagu commented:
@samuelcolvin Okay, I think I addressed the points you raised. I slightly reworked where which exceptions arise in order to make the messages a little more clear. Let me know if you think it needs more work.

dmontagu commented on pydantic/generics.py:
```diff
@@ -0,0 +1,85 @@
+            params = (params,)
+        if any(isinstance(param, TypeVar) for param in params):  # type: ignore
+            raise TypeError(f'Type parameters should be placed on typing.Generic, not GenericModel')
+        if Generic not in cls.__bases__:
```
It's important to put this check for inheriting from Generic *after* the check for TypeVar parameterization, or the error messages end up being more confusing if you try to put type vars directly on the GenericModel (which I would expect to be a relatively common mistake).

intrepidOlivia commented:
Do we have an ETA on when this might be merged?

drunkwcodes commented:
I'm also curious about it but I still stand for @dmontagu's optimistic development cycle.

samuelcolvin commented:
this is amazing, thank you very much.

samuelcolvin commented:
@intrepidOlivia @drunkwcodes I've released a new version now with this included, should be available in pypi in a few minutes.

intrepidOlivia commented:
Awesome! Thank you!
